packages/components/src/utils/hooks/stories/use-cx.js

```diff
@@ -65,6 +70,50 @@ const Example = ( { args, children } ) => {
 	return <span className={ classes }>{ children }</span>;
 };
 
+export const _slotfill = () => {
+	const { Fill, Slot } = createSlotFill( 'ToolsPanelSlot' );
+
+	const redText = css`
+		color: red;
+	`;
+	const blueText = css`
+		color: blue;
+	`;
+	const greenText = css`
+		color: green;
+	`;
+
+	return (
+		<SlotFillProvider>
+			<StyleProvider document={ document }>
+				<IFrame>
+					<IFrame>
+						<Example args={ [ redText ] }>
+							This text is inside an iframe and should be red
+						</Example>
+						<Fill name="test-slot">
+							<Example args={ [ blueText ] }>
+								This text is also inside the iframe, but is
+								relocated by a slot/fill and should be blue
+							</Example>
+						</Fill>
+						<Fill name="outside-frame">
+							<Example args={ [ greenText ] }>
+								This text is also inside the iframe, but is
+								relocated by a slot/fill and should be green
+							</Example>
+						</Fill>
+					</IFrame>
+					<StyleProvider document={ document }>
+						<Slot name="test-slot" />
+					</StyleProvider>
+				</IFrame>
+				<Slot name="outside-frame" />
```

Comment:
can you try `bubblesVirtually` prop here?

Reply:
Tried, but it doesn't seem to make a difference

<img width="977" alt="Screenshot 2022-01-26 at 12 26 48" src="https://user-images.githubusercontent.com/1083581/151155277-23e10764-59ac-40ec-9376-79ae4b6f95a6.png">

Reply:
My fix for sure is specific to "bubblesVirtually" implementation which is the one we prefer in Gutenberg in general. (the other is more "historic"). I don't know why it's not working though in storybook, the  blue text is the exact same use case as spinner it seems 🤔

Reply:
This is slightly different though as in the story we've wrapped the `Slot` in a StyleProvider as well. Are we doing that in usage in Gutenberg?

This is why I was confused because I was actually able to solve the problem with the `<StyleProvider><Slot /></StyleProvider>` with the context based code that @ciampo and I came up with that I shared here: https://github.com/WordPress/gutenberg/pull/37551#issuecomment-1021447478

But it didn't work to fix the spinner.

Which makes me thing that this story, as constructed currently, while it raises an edge case that we need to cover, is actually a red herring when compared to the spinner case (and probably the preview iframe issue you were having before @youknowriad).

In short: I don't believe we've been able to actually reproduce the issue that was being found in Gutenberg in the story yet.

If I'm understanding the fix for the GB issue correctly, the minimal reproduction would just be:

```js
<Iframe>
  <Fill name="test"><Example /></Fill>
</Iframe>
<Slot name="test" />
```

And indeed, that is what appears to happen in Gutenberg. But in the story, when you try this, it worked perfectly fine.

It's just occurred to me that potentially it could be because the story is running in an iframe? I didn't get around to trying that minimal reproduction story popped out into it's own document so maybe that would make it possible to reproduce (though should theoretically not matter if we consider that any iframe can just be the "root" document of the page, it doesn't have to be the root document of the window).

Who knows!

Reply:
I think I managed to work out a Storybook examples which breaks before this fix, and works after this fix.

In an effort to replicate as closely as possible the issue that we're trying to reproduce:

- I replaced the local `IFrame` component with the `Iframe` component from `@wordpress/block-editor`
- I added `bubblesVirtually` to both `<Slot>`s

| without the fix in this PR | with the fix in this PR, but without `bubblesVirtually` | with the fix in this PR, and with `bubblesVirtually` |
|---|---|---|
| <img width="650" alt="Screenshot 2022-01-26 at 18 48 18" src="https://user-images.githubusercontent.com/1083581/151218283-affb2ed2-893a-4f45-915f-606bd75c891c.png"> | <img width="546" alt="Screenshot 2022-01-26 at 18 55 54" src="https://user-images.githubusercontent.com/1083581/151219410-b2759855-49f5-4908-a2a1-7947df98ce38.png"> |  <img width="719" alt="Screenshot 2022-01-26 at 18 50 32" src="https://user-images.githubusercontent.com/1083581/151218620-324b3983-b498-4c61-8281-e9ab9378a030.png"> |

A couple more notes:

- changing the `iframe` component seems to be what "allowed the example to be broken"
- adding `bubblesVirtually` allowed the `slot` to receive the fix from this PR

I guess an interesting thing would be to understand what was the difference between the two `IFrame` components?

Reply:
> I guess an interesting thing would be to understand what was the difference between the two IFrame components?

This would be good to answer to prevent future changes to the component breaking things. The original one from the storybook was a copy/paste from the version from the block-editor at the time but with a lot of things removed that seemed specific to Gutenberg. I guess that might have been a mistake in the end as far as testability goes!

@ciampo is it possible to write a unit test for this btw? I'm not sure how JSDom works with iframes, if it has the same implications for style application 🤔

Reply:
I've also added [the simpler "minimal reproduction" example](https://github.com/WordPress/gutenberg/pull/38237/commits/686e8483e9b75e241a86e21e4f148e621e13ea2f) suggested by @sarayourfriend in a previous message, which with the recent changes (different `IFrame` component and `bubblesVirtually` prop) also breaks without the fix in this PR, and works after applying the fix in this PR

| without fix | with fix |
|---|---|
| <img width="336" alt="Screenshot 2022-01-26 at 19 06 20" src="https://user-images.githubusercontent.com/1083581/151221049-9e76a6d2-84b3-4a9b-aab7-16bf963b03f7.png"> | <img width="346" alt="Screenshot 2022-01-26 at 19 05 58" src="https://user-images.githubusercontent.com/1083581/151220973-1f3b1c71-218f-4079-b5d6-6d472a6238ec.png"> |

Reply:
> is it possible to write a unit test for this btw? I'm not sure how JSDom works with iframes, if it has the same implications for style application 🤔

I'll give myself tomorrow to try and write one. In case I don't manage, I guess we can then go ahead and merge this PR to unblock #37551.

Separately to this, @sarayourfriend do you think that it's worth investigating the fix you proposed in https://github.com/WordPress/gutenberg/pull/37551#issuecomment-1021447478 in a separate issue/PR?

It would also be good to understand if any of these fixes would also fix the issue highlighted in https://github.com/WordPress/gutenberg/pull/35619#discussion_r783270274

Reply:
I spent a good portion of time trying to write a test but, as feared, I didn't manage to put together a good unit test.

Apart from various minor complications, the main issue seems to be the with the `IFrame` component — it looks like [`iframeDocument`](https://github.com/WordPress/gutenberg/blob/f2161e246b9fdd9a2a56e7552b0b28050f1a5302/packages/block-editor/src/components/iframe/index.js#L269) is not defined when running the tests, and therefore the contents of the `IFrame` never render.

I'm going to give it a final look, then polish these changes and merge. 

> I have a feeling that the fix from #37551 (comment) might be relevant to #35619 (comment) in a way that the fix in this current PR doesn't seem to be. I will try that out tomorrow.

🧐 You may be onto something there. Please do keep me up to date with your progress!

Reply:
> Apart from various minor complications, the main issue seems to be the with the IFrame component — it looks like iframeDocument is not defined when running the tests, and therefore the contents of the IFrame never render.

Maybe there's some combination of `waitFor`'s that would resolve that. It could also be related to the `load` event not firing in jsdom, preventing it from ever appearing? 🤷 JSDom is such a funny thing. Most of the time it's a fine stand-in for the real thing but every so often, in particular with weird edge cases like this, it just doesn't simulate _enough_ of a real DOM to fulfill the needs for unit testing.

An e2e test written against the storybook would be interesting, but I don't believe there's any infrastructure for that at the moment in Gutenberg.

Reply:
I managed to get around the `IFrame` loading with a bit of a hacky approach as nothing else seemed to work me, but the problem seems to be related to the contents of `createPortal` not being rendered (or maybe simply "seen"?) by Jest / RTL.

These are the changes that I made so far (I've also added some "debugging" spans to the `Iframe` component to help with troubleshooting):

<details>
<summary>Click to see diff</summary>

```diff
diff --git a/packages/block-editor/src/components/iframe/index.js b/packages/block-editor/src/components/iframe/index.js
index b4e1e44f65..d9cc27a1be 100644
--- a/packages/block-editor/src/components/iframe/index.js
+++ b/packages/block-editor/src/components/iframe/index.js
@@ -267,23 +267,33 @@ function Iframe(
 				title={ __( 'Editor canvas' ) }
 			>
 				{ iframeDocument &&
-					createPortal(
-						<>
-							<head ref={ headRef }>{ head }</head>
-							<body
-								ref={ bodyRef }
-								className={ classnames(
-									BODY_CLASS_NAME,
-									...bodyClasses
+					( () => {
+						return (
+							<>
+								<span>Outside the portal</span>
+								{ createPortal(
+									<>
+										<head ref={ headRef }>{ head }</head>
+										<body
+											ref={ bodyRef }
+											className={ classnames(
+												BODY_CLASS_NAME,
+												...bodyClasses
+											) }
+										>
+											<StyleProvider
+												document={ iframeDocument }
+											>
+												<span>Inside the portal!</span>
+												{ children }
+											</StyleProvider>
+										</body>
+									</>,
+									iframeDocument.documentElement
 								) }
-							>
-								<StyleProvider document={ iframeDocument }>
-									{ children }
-								</StyleProvider>
-							</body>
-						</>,
-						iframeDocument.documentElement
-					) }
+							</>
+						);
+					} )() }
 			</iframe>
 			{ tabIndex >= 0 && after }
 		</>
diff --git a/packages/components/src/utils/hooks/test/use-cx.js b/packages/components/src/utils/hooks/test/use-cx.js
index 181e0beeca..947158f0ad 100644
--- a/packages/components/src/utils/hooks/test/use-cx.js
+++ b/packages/components/src/utils/hooks/test/use-cx.js
@@ -4,14 +4,29 @@
 // eslint-disable-next-line no-restricted-imports
 import { cx as innerCx } from '@emotion/css';
 import { insertStyles } from '@emotion/utils';
-import { render } from '@testing-library/react';
+import {
+	screen,
+	render,
+	waitFor,
+	waitForElementToBeRemoved,
+} from '@testing-library/react';
 import { css, CacheProvider } from '@emotion/react';
 import createCache from '@emotion/cache';
 
+/**
+ * WordPress dependencies
+ */
+import { __unstableIframe as Iframe } from '@wordpress/block-editor';
+import { useState } from '@wordpress/element';
+
 /**
  * Internal dependencies
  */
 import { useCx } from '..';
+import {
+	createSlotFill,
+	Provider as SlotFillProvider,
+} from '../../../slot-fill';
 
 jest.mock( '@emotion/css', () => ( {
 	cx: jest.fn(),
@@ -61,4 +76,65 @@ describe( 'useCx', () => {
 			false
 		);
 	} );
+
+	it( 'should work correctly when using slot/fill in combination with an IFrame', async () => {
+		// const { Fill, Slot } = createSlotFill( 'UseCxTest' );
+
+		// const redText = css`
+		// 	color: red;
+		// `;
+
+		// const TestComponent = () => {
+		// 	const [ iframeIsLoading, setIframeIsLoading ] = useState( true );
+
+		// 	return (
+		// 		<SlotFillProvider>
+		// 			{ iframeIsLoading && (
+		// 				<span data-testid="iframe-loading">Loading</span>
+		// 			) }
+		// 			<Iframe onLoad={ () => setIframeIsLoading( false ) }>
+		// 				<span data-testid="iframe-simple-content">
+		// 					Iframe Content
+		// 				</span>
+		// 				<Fill name="test-slot">
+		// 					<Example args={ [ redText ] }>
+		// 						This text should be red
+		// 					</Example>
+		// 				</Fill>
+		// 			</Iframe>
+		// 			<Slot bubblesVirtually name="test-slot" />
+		// 		</SlotFillProvider>
+		// 	);
+		// };
+
+		const SimplifiedTestComponent = () => {
+			const [ iframeIsLoading, setIframeIsLoading ] = useState( true );
+			return (
+				<div>
+					{ iframeIsLoading && (
+						<span data-testid="iframe-loading">Loading</span>
+					) }
+					<Iframe onLoad={ () => setIframeIsLoading( false ) }>
+						<span data-testid="iframe-simple-content">
+							Iframe Content
+						</span>
+					</Iframe>
+				</div>
+			);
+		};
+
+		const { debug } = render( <SimplifiedTestComponent /> );
+
+		let loadingSpy;
+		await waitFor( () => {
+			loadingSpy = screen.getByTestId( 'iframe-loading' );
+			expect( loadingSpy ).toBeInTheDocument();
+		} );
+
+		await waitFor( () => expect( loadingSpy ).not.toBeInTheDocument() );
+
+		// <span>Outside the portal</span> is printed out
+		// <span>Inside the portal!</span> is NOT printed out (together with the rest of the children)
+		debug();
+	} );
 } );
```
</details>

If anyone has the time to take a look as well it'd be great! Otherwise I plan on spending a little bit more time on it tomorrow.

Reply:
Just checking if you tried the suggestion in this issue? https://github.com/testing-library/react-testing-library/issues/62

I never really ran into any problems testing portals as far as I can remember but I might have just gotten lucky and avoided testing them directly so far!

Seems like the trick is to pass an extra `container: document.body` argument to `render`.

Reply:
Unfortunately adding `{ container: document.body }` as an additional argument to `render`  triggers a warning, and the `renderIntoDocument` method (also mentioned in that issue) is deprecated.

I've tried to get inspiration from [this example](https://github.com/kentcdodds/react-testing-library-examples/blob/main/src/__tests__/portals.js) but as soon as the portal doesn't render in the `iframeDocument.documentElement`, more errors appear (for example, [`ownerNode` here](https://github.com/WordPress/gutenberg/blob/f2161e246b9fdd9a2a56e7552b0b28050f1a5302/packages/block-editor/src/components/iframe/index.js#L59) becomes `undefined`).

Since I've almost spent 2 days on this without managing to get a working unit test, I'm going to merge this PR as-is in order to unblock #37551

Reply:
> I have a feeling that the fix from #37551 (comment) might be relevant to #35619 (comment) in a way that the fix in this current PR doesn't seem to be. I will try that out tomorrow.

Did you manage to have a look at this at all? If not, I can try to also look into it later today / on Monday.

Reply:
I was away from my computer yesterday so I didn't get to, but if I have time to look into it today I will!